test/00_diff/BasicDiffTests.cu

```diff
@@ -0,0 +1,204 @@
+  }
+
+  (t1o = deriv(pow(wrt(t1i),TypeParam(2)))).run(stream);
+  cudaStreamSynchronize(stream);
+  for (index_t i = 0; i <t1i.Size(0); i++) {
+     auto x = t1i(i);
+     ASSERT_NEAR(t1o(i), TypeParam(2)*x, .001*scale);
+  }
+
+  (t1o = deriv(pow(TypeParam(2),wrt(t1i)))).run(stream);
+  cudaStreamSynchronize(stream);
+  for (index_t i = 0; i <t1i.Size(0); i++) {
+     auto x = t1i(i);
+     ASSERT_NEAR(t1o(i), log(TypeParam(2))*pow(TypeParam(2),x),.001*scale);
+  }
+
+  (t1o = deriv(pow(wrt(t1i),wrt(t1i)))).run(stream);
+  cudaStreamSynchronize(stream);
+  for (index_t i = 0; i <t1i.Size(0); i++) {
+     auto x = t1i(i);
+     ASSERT_FLOAT_EQ(t1o(i), (log(x)+TypeParam(1))*pow(x,x));
+  }
+  
+  (t1o = deriv(pow(sin(wrt(t1i)),TypeParam(2)))).run(stream);
+  cudaStreamSynchronize(stream);
+  for (index_t i = 0; i <t1i.Size(0); i++) {
+     auto x = t1i(i);
+     ASSERT_NEAR(t1o(i), TypeParam(2)*cos(x)*sin(x),.0001*scale);
+  }
+  
+  (t1o = deriv(pow(wrt(t1i),sin(wrt(t1i))))).run(stream);
+  cudaStreamSynchronize(stream);
+  for (index_t i = 0; i <t1i.Size(0); i++) {
+     auto x = t1i(i);
+     auto s = sin(x);
+     auto c = cos(x);
+     auto l = log(x);
+
+     ASSERT_NEAR(t1o(i), pow(x,sin(x)) * ( s/x + c*l ),.0001*scale);
+  }
+  
+  auto y = TypeParam(1) + sin(wrt(t1i)) - sin(wrt(t1i));
+  (t1o = deriv(pow(y,y))).run(stream);
+  cudaStreamSynchronize(stream);
+  for (index_t i = 0; i <t1i.Size(0); i++) {
+     auto x = t1i(i);
+     ASSERT_NEAR(t1o(i), TypeParam(0) ,.0001*scale);
+  }
+
+  (t1o = deriv(pow(sin(wrt(t1i))-sin(wrt(t1i))+TypeParam(1),sin(wrt(t1i))))).run(stream);
+  cudaStreamSynchronize(stream);
+  for (index_t i = 0; i <t1i.Size(0); i++) {
+    
+    auto x = t1i(i);
+    auto c = cos(x);
+    auto s = sin(x);
+    auto v = c*pow(s,s)*(log(s))+TypeParam(1);
+
+    ASSERT_NEAR(t1o(i), TypeParam(0),.0001*scale);
+  }
+
+  (t1o = deriv(sin(wrt(t1i)))).run(stream);
+  cudaStreamSynchronize(stream);
+  for (index_t i = 0; i <t1i.Size(0); i++) {
+     auto x = t1i(i);
+     ASSERT_FLOAT_EQ(t1o(i), cos(x));
+  }
+
+  (t1o = deriv(sin(sin(wrt(t1i))))).run(stream);
+  cudaStreamSynchronize(stream);
+  for (index_t i = 0; i <t1i.Size(0); i++) {
+     auto x = t1i(i);
+     ASSERT_NEAR(t1o(i), cos(x)*cos(sin(x)),.00001*scale);
+  }
+  (t1o = deriv(sin(wrt(t1i))+cos(wrt(t1i)))).run(stream);
+  cudaStreamSynchronize(stream);
+  for (index_t i = 0; i <t1i.Size(0); i++) {
+     ASSERT_NEAR(t1o(i), cos(t1i(i))-sin(t1i(i)),.00001*scale);
+  }
+  (t1o = deriv(sin(wrt(t1i))-cos(wrt(t1i)))).run(stream);
+  cudaStreamSynchronize(stream);
+  for (index_t i = 0; i <t1i.Size(0); i++) {
+     ASSERT_NEAR(t1o(i), cos(t1i(i))+sin(t1i(i)),.00001*scale);
+  }
+  
+  (t1o = deriv(sin(wrt(t1i))*cos(wrt(t1i)))).run(stream);
+  cudaStreamSynchronize(stream);
+  for (index_t i = 0; i <t1i.Size(0); i++) {
+     ASSERT_NEAR(t1o(i), cos(t1i(i))*cos(t1i(i))-sin(t1i(i))*sin(t1i(i)), .00001*scale);
+  }
+ 
+  if(!is_matx_half_v<TypeParam> ) { 
```

Comment:
Why is half removed here? If it's the ambiguous cast can you cast it to a float if it's a half type?

Reply:
half precision does not have enough dynamic range to resolve the solution in any meaningful range.  It compiles fine it just doesn't provide a useful answer.   I'll add a comment to clarify.